// @game/game-framework-laya/module/framework/ui/BottomMenu.js
import LoggerUtil from '../util/LoggerUtil';
/**
 * 底部菜单父类
 */
export default class BottomMenu extends Laya.View {
    constructor() {
        super();
        this._isCloseOnSide = true;
    }
    /**
     * logan日志上报名字
     * 一般使用类名
     * 需要子类在构造方法中调用
     * @param loganName
     */
    setLoganName(loganName) {
        this.loganName = loganName;
    }
    /**
     * 设置是否响应点击底部菜单外部区域关闭
     * @param isCloseOnSide 是否影响
     */
    setCloseOnSide(isCloseOnSide) {
        this._isCloseOnSide = isCloseOnSide;
    }
    /**
     * 返回是否响应点击底部菜单外部区域关闭
     */
    isCloseOnSide() {
        return this._isCloseOnSide;
    }
    /**
     * 设置点击底部菜单外部区域执行方法
     * @param closeOnSideClickListener 点击弹框外部区域执行方法
     */
    setCloseOnSideClickListener(closeOnSideClickListener) {
        this._closeOnSideClickListener = closeOnSideClickListener;
    }
    /**
     * 返回已设置点击底部菜单外部区域执行方法
     */
    getCloseOnSideClickListener() {
        return this._closeOnSideClickListener;
    }
    /**
     * 关闭底部菜单
     */
    close() {
        this.menuView.close();
    }
    /**
     * [游戏框架内部调用]
     * 设置menuView
     * @param menuView
     */
    setMenuView(menuView) {
        this.menuView = menuView;
    }
    onOpened() {
        if (this.loganName) {
            LoggerUtil.getInstance().info('[BottomMenu] onOpened', this.loganName);
        }
        else {
            LoggerUtil.getInstance().warn('[BottomMenu] onOpened loganName error!', this.constructor.name);
        }
    }
    onClosed() {
        if (this.loganName) {
            LoggerUtil.getInstance().info('[BottomMenu] onClosed', this.loganName);
        }
        else {
            LoggerUtil.getInstance().warn('[BottomMenu] onClosed loganName error!', this.constructor.name);
        }
        this.destroy();
    }
}
