// @game/game-framework-laya/module/framework/ui/GUIManager.js
import TouchBlockLayer from './TouchBlockLayer';
import BottomMenuView from './BottomMenuView';
import Singleton from '../util/Singleton';
import GlobalZorder from './GlobalZorder';
import ButtonScale from './ButtonScale';
import AppBridge from '../bridge/AppBridge';
import DeviceUtil from '../util/DeviceUtil';
/**
 * GUI管理器
 * 处理其他GUI的弹出关闭逻辑
 */
export default class GUIManager extends Singleton {
    /**
     * 显示全局的触摸屏蔽层
     * @param isGray
     */
    showTouchBlockLayer(isGray) {
        if (!this.touchBlockLayer) {
            this.touchBlockLayer = new TouchBlockLayer(isGray);
            Laya.stage.addChild(this.touchBlockLayer);
        }
    }
    /**
     * 返回全局的触摸屏蔽层
     */
    getTouchBlockLayer() {
        return this.touchBlockLayer;
    }
    /**
     * 移除全局的触摸屏蔽层
     */
    closeTouchBlockLayer() {
        if (this.touchBlockLayer) {
            this.touchBlockLayer.destroy();
            this.touchBlockLayer = null;
        }
    }
    /**
     * 显示底部菜单
     * @param bottomMenu 底部菜单
     */
    showBottomMenu(bottomMenu) {
        this.closeBottomMenu();
        let menuView = new BottomMenuView(bottomMenu);
        this.bottomMenu = bottomMenu;
        Laya.stage.addChild(menuView);
    }
    /**
     * 返回底部菜单
     */
    getBottomMenu() {
        if (this.bottomMenu) {
            return this.bottomMenu;
        }
    }
    /**
     * 关闭底部菜单
     */
    closeBottomMenu() {
        if (this.bottomMenu) {
            this.bottomMenu.close();
            this.bottomMenu = null;
        }
    }
    /**
     * 显示全局关闭按钮
     */
    showGlobalCloseBtn(globalCloseBtnParam) {
        if (!globalCloseBtnParam) {
            globalCloseBtnParam = {};
        }
        if (globalCloseBtnParam.visible === undefined) {
            globalCloseBtnParam.visible = true;
        }
        if (globalCloseBtnParam.right === undefined) {
            globalCloseBtnParam.right = 20;
        }
        if (globalCloseBtnParam.y === undefined) {
            globalCloseBtnParam.y = 26;
        }
        if (globalCloseBtnParam.skin === undefined) {
            globalCloseBtnParam.skin = 'app/common/globalCloseBtn.png';
        }
        if (!this.globalCloseBtn && globalCloseBtnParam.visible) {
            let closeBtn = new Laya.Image();
            this.globalCloseBtn = closeBtn;
            closeBtn.skin = globalCloseBtnParam.skin;
            closeBtn.zOrder = GlobalZorder.globalCloseBtn;
            closeBtn.addComponent(ButtonScale);
            Laya.stage.addChild(closeBtn);
            closeBtn.on(Laya.Event.CLICK, this, () => {
                let isExecute = AppBridge.getInstance().event('onBeforeClosePage');
                if (!isExecute) {
                    AppBridge.getInstance().exitMiniProgram();
                }
            });
            closeBtn.right = globalCloseBtnParam.right;
            closeBtn.y = DeviceUtil.getInstance().getNotchLength() + globalCloseBtnParam.y;
        }
    }
    /**
     * 返回全局关闭按钮
     * 只有web渠道可以返回
     */
    getGlobalCloseBtn() {
        return this.globalCloseBtn;
    }
}
