// @game/game-framework-laya/lib/framework/init.ts
import { BridgeInterface } from './bridge/BridgeInterface';
import { LoginSDKInterface } from './login/LoginSDKInterface';
import RainbowLauncher from './rainbow/RainbowLauncher';

/**
 * 游戏框架初始化
 *
 */

/**
 * 游戏框架初始化
 * @param option
 * requestPortmConfig（后续会被废弃，请尽快迁移至使用requestMgcConfig） 从portm上拉取强更配置和分享配置，历史默认值为true（历史原因），替代方案为使用requestMgcConfig
 * requestMgcConfig 从mgcconfig上拉取强更配置和分享配置，如果已经将porm配置迁移至mgcconfig，请将该值设置为true，否则设置为false；
 *    当设置为true时会忽略requestPortmConfig的值（不会出现既从mgcconfig，又从portm上拉取强更配置和分享配置的情况）
 */
export function framework_init(option: {
  gameName: string;
  width: number;
  height: number;
  suggestAspectRatio?: number;
  bridge?: BridgeInterface;
  loginSDK?: LoginSDKInterface;
  globalCloseBtn?: {
    visible?: boolean;
    right?: number;
    y?: number;
    skin?: string;
  };
  bgColor?: string;
  defaultFontSize?: number;
  defaultFont?: string;
  defaultMaskAlpha?: number;
  scaleMode?: string;
  hideLaunchScreenDelayTime?: number;
  useRetinalCanvas?: boolean;
  useWebGL2?: boolean;
  dialogShowEffect?: Laya.Handler;
  dialogCloseEffect?: Laya.Handler;
  requestPortmConfig: boolean;
  requestMgcConfig: boolean;
}) {
  RainbowLauncher.getInstance().init(option);
}
