// @game/game-framework-laya/module/index.js
import AppBridge from './framework/bridge/AppBridge';
import DataDevice from './framework/data/DataDevice';
import DataMgcConfig from './framework/data/DataMgcConfig';
import DataManager from './framework/data/DataManager';
import LoginSDK from './framework/login/LoginSDK';
import HttpRequest from './framework/network/HttpRequest';
import EventDispatcher from './framework/message/EventDispatcher';
import DeviceUtil from './framework/util/DeviceUtil';
import AnalyticsUtil from './framework/util/AnalyticsUtil';
import Singleton from './framework/util/Singleton';
import Timestamp from './framework/util/Timestamp';
import UrlUtil from './framework/util/UrlUtil';
import VersionUtil from './framework/util/VersionUtil';
import LoggerUtil from './framework/util/LoggerUtil';
import MetricsUtil from './framework/util/MetricsUtil';
import ConfigUtil from './framework/util/ConfigUtil';
import RainbowLauncher from './framework/rainbow/RainbowLauncher';
import RainbowMessageType from './framework/rainbow/RainbowMessageType';
import RainbowConfig from './framework/rainbow/RainbowConfig';
// Bridge导出
export { BridgeType, ShareType } from './framework/bridge/BridgeInterface';
export { AppBridge };
// 数据类导出
export { DataBase, DataParseType } from './framework/data/DataBase';
export { DataDevice, DataMgcConfig };
export { DataManager };
// UI类导出
export * from './framework/ui/index';
// 登录类导出
export { LoginType } from './framework/login/LoginSDKInterface';
export { LoginSDK };
// message类导出
export { EventDispatcher };
// NETWORK类导出
export { HttpRequest };
// 调试类导出
export { framework_init } from './framework/init';
// Util类导出
export { AnalyticsUtil };
export { DeviceUtil };
export * from './framework/util/functions';
export { Singleton };
export { Timestamp };
export { UrlUtil };
export { VersionUtil };
export { LoggerUtil };
export { MetricsUtil };
export { ConfigUtil };
export { RainbowLauncher, RainbowMessageType, RainbowConfig };
