// @game/game-framework-laya/lib/framework/ui/SceneManager.ts
import TouchBlockLayer from './TouchBlockLayer';
import GlobalZorder from './GlobalZorder';
import SceneBase from './SceneBase';
import Singleton from '../util/Singleton';

/**
 * 界面管理器
 */
export default class SceneManager extends Singleton {
  private scene: SceneBase;

  private preScene: SceneBase;

  private touchBlockLayer: TouchBlockLayer;

  /**
   * 显示界面
   * @param scene 界面对象
   */
  public showScene(scene: SceneBase): void {
    //Laya.Dialog.manager.closeAll();
    if (this.preScene) {
      this.preScene.destroy();
      this.preScene = null;
    }

    this.preScene = this.scene;
    this.scene = scene;
    scene.zOrder = GlobalZorder.scene;
    Laya.stage.addChild(scene);

    if (this.preScene) {
      this.scene.visible = false;
      if (!this.touchBlockLayer) {
        this.touchBlockLayer = new TouchBlockLayer(true);
        Laya.stage.addChild(this.touchBlockLayer);
      }

      this.preScene.doClose();

      Laya.Tween.to(
        this.touchBlockLayer,
        { alpha: 0.8 },
        100,
        null,
        new Laya.Handler(this, () => {
          if(this.preScene){
            this.preScene.onClosed();
            this.preScene.destroy();
            this.preScene = null;
          }
          this.scene.visible = true;
          Laya.Tween.to(
            this.touchBlockLayer,
            { alpha: 0 },
            100,
            null,
            new Laya.Handler(this, () => {
              this.touchBlockLayer.destroy();
              this.touchBlockLayer = null;

              this.scene.onOpened();
            })
          );
        })
      );
    } else {
      Laya.timer.once(1, this, () => {
        this.scene.onOpened();
      });
    }
  }

  /**
   * 返回当前界面
   */
  public getScene(): SceneBase {
    return this.scene;
  }
}
