// @game/game-framework-laya/module/framework/util/DeviceUtil.js
import Singleton from './Singleton';
/**
 * 设备处理类
 */
export default class DeviceUtil extends Singleton {
    constructor() {
        super(...arguments);
        this.notchLength = 60;
    }
    /**
     * 是否有刘海
     * 长宽比>1.99的手机都认为有刘海
     */
    hasNotch() {
        if (Laya.stage.height / Laya.stage.width >= this.getNotchLengthWidthRate()) {
            return true;
        }
        return false;
    }
    /**
     * 设置刘海高度
     * @param notchLength 刘海高度，默认为60
     */
    setDefaultNotchLength(notchLength) {
        this.notchLength = notchLength;
    }
    /**
     * 返回刘海高度
     */
    getNotchLength() {
        if (Laya.stage.width > Laya.stage.height) {
            if (Laya.stage.width / Laya.stage.height >= this.getNotchLengthWidthRate()) {
                return this.notchLength;
            }
        }
        else {
            if (Laya.stage.height / Laya.stage.width >= this.getNotchLengthWidthRate()) {
                return this.notchLength;
            }
        }
        return 0;
    }
    getNotchLengthWidthRate() {
        if (process.env.miniprogram) {
            return 1.8;
        }
        return 1.99;
    }
}
